File: app/src/main/java/tech/pauly/findapet/utils/FilterChipListLayout.java
package tech.pauly.findapet.utils;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import tech.pauly.findapet.R;
import tech.pauly.findapet.discover.Chip;

public class FilterChipListLayout extends LinearLayout {

    private Context context;
    @Nullable
    private Chip locationChip;
    private List<Chip> filterChips = new ArrayList<>();

    public FilterChipListLayout(Context context) {
        super(context);
        setup(context);
    }

    public FilterChipListLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setup(context);
    }

    public FilterChipListLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setup(context);
    }

    private void setup(Context context) {
        this.context = context;
        updateLayout();
    }

    public void setLocationChip(Chip chip) {
        this.locationChip = chip;
        updateLayout();
    }

    public void setFilterChips(List<Chip> filterChips) {
        this.filterChips = filterChips;
        updateLayout();
    }

    private void updateLayout() {
        if (filterChips.size() > 0 || locationChip != null) {
            setVisibility(VISIBLE);
            removeAllViews();
            LayoutInflater inflater = LayoutInflater.from(context);
            if (locationChip != null) {
                addChip(inflater, locationChip);
            }
            for (Chip chip : filterChips) {
                addChip(inflater, chip);
            }
        } else {
            setVisibility(GONE);
        }
        requestLayout();
    }

    private void addChip(LayoutInflater inflater, Chip chip) {
        FrameLayout chipLayout = (FrameLayout) inflater.inflate(R.layout.item_chip, this, false);
        ((TextView) chipLayout.getChildAt(0)).setText(chip.getText());
        addView(chipLayout);
    }
}

File: app/src/main/java/tech/pauly/findapet/utils/FilterChipListLayout.kt
package tech.pauly.findapet.utils

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import tech.pauly.findapet.R
import java.util.*

data class Chip(val text: String)

class FilterChipListLayout : LinearLayout {

    private var locationChip: Chip? = null
    private var filterChips: List<Chip> = ArrayList()

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        updateLayout()
    }

    fun setLocationChip(chip: Chip?) {
        this.locationChip = chip
        updateLayout()
    }

    fun setFilterChips(filterChips: List<Chip>) {
        this.filterChips = filterChips
        updateLayout()
    }

    private fun updateLayout() {
        removeAllViews()
        val inflater = LayoutInflater.from(context)
        locationChip?.also { addChip(inflater, it) }
        filterChips.forEach { addChip(inflater, it) }
        visibility = if (childCount == 0) View.GONE else View.VISIBLE
        requestLayout()
    }

    private fun addChip(inflater: LayoutInflater, chip: Chip) {
        (inflater.inflate(R.layout.item_chip, this, false) as FrameLayout).let {
            (it.getChildAt(0) as? TextView)?.text = chip.text
            addView(it)
        }
    }
}